### PR TITLE
Threading issue in sf::SoundRecorder

### PR DESCRIPTION
**Pull request by Foaly**
When you destroy a `sf::SoundBufferRecorder` while it's still recording you get the following crash
```
pure virtual method called
terminate called without an active exception
```
This is because the `sf::SoundBufferRecorder::onProcessSamples()` method gets deleted while the capturing thread in `sf::SoundRecorder` i still running and calling it (destructor of derived class
get called first). Simply calling `stop()` in the destructor fixes it. 

This reproduces the crash:
```cpp
int main() {
    // create the recorder
    sf::SoundBufferRecorder recorder;

    // start the capture
    recorder.start();

    return 0;
}
```

Once this is merged I will write a little section for the tutorials metioning that you have to call `stop()` in the destructor of the class deriving from `sf::SoundRecorder`.

### CODE REVIEW

**Comment by Bromeon** [comment]
> Once this is merged I will write a little section for the tutorials metioning that you have to call `stop()` in the destructor of the class deriving from `sf::SoundRecorder`.

Why would the user need to do this manually? The base class destructor is called automatically.

---- 

Not really related to this PR, but I noticed that `m_isCapturing` is used from two threads without a mutex, which is not ideal...

**Comment by eXpl0it3r** [comment]
I think you only need to call `stop()` in the destructor of `SoundRecorder`, since `SoundBufferRecorder` derives from `SoundRecorder` so you'd be calling `stop()` twice.
And neither do you have to call `stop()` in the example code.

**Comment by Foaly** [comment]
This is not entire correct. The problem here is timing. It is true that the base class constructor is being called when the derived object is destoryed, but between those two calls the capture thread could be actived and try to call a deleted method, which leads to a crash.

Let me explain in detail. Right now it works like this: `SoundBufferRecorder::onProcessSamples` overwrittes the pure virtual function `SoundRecorder::onProcessSamples` which is called by the audio capture thread method. When `SoundBufferRecorder` destructor is called it's methods get deleted. The capturing thread however is still active, since noone called `stop()`. It the tryes to call `SoundBufferRecorder::onProcessSamples`, which is deleted and results in the `pure virtual function called` crashed.
Putting `stop()` in the base class destructor doesn't solve the problem entirely, because the base class destructor is called after the derived classes destructor. So if the capturing thread is activated after the methods have been deleted by the derived destructor, but before it is stopped by the base class destructor you have the same crash again.
You can try this out by removing `stop()` [here](https://github.com/SFML/SFML/pull/1011/files#diff-a574f234d93c4d82fa0801c77b4df053R39). Now the program will only crash sometimes (unpredictable, scheduler dependant). If you replace that line with `sf::sleep(sf::milliseconds(110));` (m_processingInterval + 10 to make sure the thread function is at least called once) the crash happens everytime.
So the `stop()` call in the derived calls destructor is nesseccary.

**tl;dr**: The `stop()` call in the destructor of the derived class is nesseccary, because the capturing thread has to be stopped before the methods of the derived class (which the threads uses) are deleted.

**Comment by eXpl0it3r** [comment]
I see. Question then is, should the base class still call `stop()`, even though every derived class is required to call it? Seems to me it could potentially hide a missing `stop()` in the derived class with a race condition.

**Comment by Foaly** [comment]
Well that decision is up to you guys I guess. You could leave it in, so even if the user forgets to call stop in the destructor of the derived class, the whole program doesn't crash right away. Chances that the thread is executed between the two destructor calls are pretty slim. But then again you could argue that that hides a programming error and that you rather have a crash to enforce the correct programming on the user.

**Comment by Bromeon** [comment]
> When `SoundBufferRecorder` destructor is called it's methods get deleted.

The methods themselves are not destroyed, the object and its VTable are; but you're right in that the access to virtual functions and member variables from the thread `SoundRecorder::record()` are dangerous.

Good observation! This explanation should be part of the docs, because it's really not obvious why the base class destructor can't do it.

> You could leave it in, so even if the user forgets to call stop in the destructor of the derived class, the whole program doesn't crash right away.

That's a sneaky way of obscuring application-breaking problems. **DO NOT HIDE LOGIC ERRORS!**
I can only repeat myself, but you don't help anybody by making a dangerous error less noticeable. What we should do is check in the base destructor whether `stop()` was called, and trigger an assertion if it wasn't.

**Comment by Foaly** [comment]
Sorry I'll try to not hide logic erros in the future. I didn't think of using an assertion. I updated the code.

> Good observation! This explanation should be part of the docs[...]

Thanks! I added some documentation to the `SoundRecorder.hpp`. Should there be more documentation or a better explanation in the code?
Like I said I wanted to write a paragraf in the [Threading issue section](http://www.sfml-dev.org/tutorials/2.3/audio-recording.php#threading-issues) of the tutorials once this is merged.

**Comment by Bromeon** [comment]
>  Should there be more documentation or a better explanation in the code?

Since it's something that users who inherit `sf::SoundRecorder` might wonder, I would add it to the documentation of the `sf::SoundRecorder` class.

> Like I said I wanted to write a paragraf in the [Threading issue section](http://www.sfml-dev.org/tutorials/2.3/audio-recording.php#threading-issues) of the tutorials once this is merged.

Good idea, too. It won't hurt to have it in both places :)

**Comment by Foaly** [comment]
Sorry I don't get it. There is already documentation in `sf::SoundRecorder` [here](https://github.com/SFML/SFML/pull/1011/files#diff-1). Do you want me to add a more detailed description on why the problem is happing there? It might be interessting for users to understand the problem.

**Comment by Bromeon** [comment]
For the docs, mentioning it there is enough, but that paragraph is confusing; the method is *not* deleted. You should rather write something like "You must call stop() in the destructor of your derived class, so that the recording thread finishes before your object is destroyed."

Maybe some additional (code comment) explanation in the `sf::SoundRecorder` base destructor could explain why `stop()` can't be called at that point, and an assertion is used instead. Just to avoid that future developers don't wonder why it was done this way.

**Comment by Foaly** [comment]
I updated the documentation. Once it's verifyed I will squash the commits.

**Comment by Foaly** [comment]
I just realized the [VoIP examples](https://github.com/SFML/SFML/blob/master/examples/voip/Client.cpp#L18) probably needs an update. I will try it out on the weekend.

**Comment by Foaly** [comment]
I updated the example. This should be ready for merging now!

File: include/SFML/Audio/SoundRecorder.hpp [hunk 1]
```diff
@@ -316,11 +316,20 @@ class SFML_AUDIO_API SoundRecorder : AlResource
 /// from this separate thread. It is important to keep this in
 /// mind, because you may have to take care of synchronization
 /// issues if you share data between threads.
+/// Another thing to bear in mind, is that you must call stop()
+/// in the destructor of your derived class, so that the recording
+/// thread finished before you object is destroyed.
```
Comment (Bromeon) [inline include/SFML/Audio/SoundRecorder.hpp]: "finishe**s**" ;)

**Comment by Bromeon** [comment]
Looks good. We have to keep in mind to fix the `m_isCapturing` race condition in a separate PR.

**Comment by Foaly** [comment]
Thanks for the review. I corrected the grammer mistake and squashed the commits. They were just seperate to make the review easier. Everything should be good to go now :+1:
I opened a new issue for the race condition #1038.

File: include/SFML/Audio/SoundRecorder.hpp [hunk 2]
```diff
@@ -316,11 +316,20 @@ class SFML_AUDIO_API SoundRecorder : AlResource
 /// from this separate thread. It is important to keep this in
 /// mind, because you may have to take care of synchronization
 /// issues if you share data between threads.
+/// Another thing to bear in mind, is that you must call stop()
```
Comment (TankOs) [inline include/SFML/Audio/SoundRecorder.hpp]: Drop comma.

File: include/SFML/Audio/SoundRecorder.hpp [hunk 3]
```diff
@@ -316,11 +316,20 @@ class SFML_AUDIO_API SoundRecorder : AlResource
...
+/// in the destructor of your derived class, so that the recording
+/// thread finishes before you object is destroyed.
```
Comment (TankOs) [inline include/SFML/Audio/SoundRecorder.hpp]: Typo; "before your" or "before the"

File: src/SFML/Audio/SoundRecorder.cpp [hunk 4]
```diff
@@ -59,7 +60,12 @@ m_deviceName        (getDefaultDevice())
 ////////////////////////////////////////////////////////////
 SoundRecorder::~SoundRecorder()
 {
-    // Nothing to do
+    // This assertion is triggered if the recording is still running while
+    // the object is destroyed. It ensures that stop() is called in the
+    // destructor of the derived class, which makes sure that the recording
+    // thread finishes before the derived object is destroyed. Otherwise a
+    // "pure virtual method called" exception is triggered.
```
Comment (TankOs) [inline src/SFML/Audio/SoundRecorder.cpp]: DRY; explanation is already fully covered in the assertion error message.

File: src/SFML/Audio/SoundRecorder.cpp [hunk 5]
```diff
@@ -114,12 +120,16 @@ bool SoundRecorder::start(unsigned int sampleRate)
...
+    {
+        // Stop the capturing thread
+        m_isCapturing = false;
+        m_thread.wait();
 
-    // Notify derived class
-    onStop();
+        // Notify derived class
+        onStop();
+    }
```
Comment (TankOs) [inline src/SFML/Audio/SoundRecorder.cpp]: The last two comments seem to be too much for me. It's clearly understandable from code what happens. ;) Also DRY, as the first comment already explains it all.

**Comment by TankOs** [comment]
Except for some minor documentation things (again, sorry for that :D): :+1: :cat2:

File: src/SFML/Audio/SoundRecorder.cpp (hunk 4 above)
Comment (Foaly) [inline src/SFML/Audio/SoundRecorder.cpp]: Its true that there is some duplication, but I think in this case it should stay in. This is error is definitly not trivial. I think the comment explains a little more detailed whats happening here any also why. It makes it easier to understand for future readers what kind of error is resolved here. If you read the discussion above this comment was specifically asked for.

File: src/SFML/Audio/SoundRecorder.cpp (hunk 5 above)
Comment (Foaly) [inline src/SFML/Audio/SoundRecorder.cpp]: I cleaned it up a bit.

**Comment by Foaly** [comment]
Thanks for review the code and the documentation. I implemented your suggestions.

**Comment by eXpl0it3r** [comment]
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

**Comment by eXpl0it3r** [comment]
Merged in 1ee6d1dbc639fbf961d3b68ce038edd15f142fcb
